### PR TITLE
Suggestions for spelling mistakes needs to be filtered much stronger

### PR DESCRIPTION
**Pull request by forki**
After reviewing elm's suggestion function I'm convinced we need to set the threshold for spelling corrections much higher.

I propose the following heuristic:

1) jarowinkler > 0.85 (high similarity) ==> accept 
2) jarowinkler < 0.7 ==> reject
3) editdistance <= 3 then ==> accept
4) otherwise reject

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
I really appreciate this change. Currently the light bulb appears virtually all the time if there is an error in code. When I see it, I assume it's useful, like "add open" or very good suggestion, but it's always some unrelated list of names.

**Comment by forki** [comment]
It's not ready yet - I will use damerau-levensthein for threshold and jaro-winkler for sorting

**Comment by cartermp** [comment]
Just updated the title to be a bit more clear as to what's going on here.

**Comment by vasily-kirichenko** [comment]
image

I believe it should suggest `CompilationRepresentationFlags` here (only).

**Comment by vasily-kirichenko** [comment]
image

"RecordExpr" should not be there IMHO.

**Comment by saul** [comment]
Once this is merged in, #2175 should be closed.

**Comment by forki** [comment]
@saul you referenced this PR here!?

**Comment by saul** [comment]
Sorry - meant #2066

**Comment by cartermp** [comment]
@saul Thanks for tagging the issue!

File: src/fsharp/ErrorResolutionHints.fs [hunk 1]
```diff
@@ -4,22 +4,45 @@
...
+    let length = unknownIdent.Length
     predictionsF()
-    |> Seq.choose (fun p -> 
-        let similarity = Internal.Utilities.EditDistance.JaroWinklerDistance unknownIdent (p.ToUpperInvariant())
-        if not useThreshold || similarity >= thresholdForSuggestions || p.Contains unknownIdent then
+    |> Seq.choose (fun p ->
+        let preditedText = p.ToUpperInvariant()
+        let similarity = Internal.Utilities.EditDistance.JaroWinklerDistance unknownIdent preditedText
+        if similarity >= highConfidenceThreshold || 
+           LongestPrefix preditedText unknownIdent > 4 ||
```
Comment (KevinRansom) [inline src/fsharp/ErrorResolutionHints.fs]: 4 and 3 are magic numbers. 

should the 3 be 'minStringLengthForThreshold '

File: src/utils/EditDistance.fs [hunk 2]
```diff
@@ -83,4 +83,48 @@ let JaroWinklerDistance s1 s2 =
...
     let p = 0.1
-    jaroScore + (l * p * (1.0 - jaroScore))
\ No newline at end of file
+    jaroScore + (l * p * (1.0 - jaroScore))
+
+/// Computes the restricted Damerau-Levenstein edit distance,
+/// also known as the "optimal string alignment" distance.
+///  - read more at https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance
+///  - Implementation taken from http://www.navision-blog.de/2008/11/01/damerau-levenshtein-distance-in-fsharp-part-ii/
+let private calcDamerauLevenshtein (a:string, b:string) =
```
Comment (KevinRansom) [inline src/utils/EditDistance.fs]: I have to say this looks like converted C++.  Not very idiomatic.

**Comment by forki** [comment]
image

^^ fixed as well

**Comment by forki** [comment]
@vasily-kirichenko  "RecordExpr" should not be suggested anymore.

**Comment by forki** [comment]
https://github.com/Microsoft/visualfsharp/pull/2175#issuecomment-270978566 is an interesting case. the error message is actually wrong (already wrong in VS2015). The CompilationRepresentationFlags is a union and not a module or namespace. So here it's clear we can't suggest the correct one since we're not in the correct error case. Trying to find the cause

**Comment by forki** [comment]
@vasily-kirichenko could please review the current state of this PR. As you can see in the changed tests A LOT of the weird cases are already gone!

**Comment by forki** [comment]
the async case was similar. and now fixed:

image

File: src/fsharp/FSComp.txt [hunk 3]
```diff
@@ -13,7 +13,7 @@ undefinedNameTypeIn,"The type '%s' is not defined in '%s'."
 undefinedNameRecordLabelOrNamespace,"The record label or namespace '%s' is not defined."
 undefinedNameRecordLabel,"The record label '%s' is not defined."
 undefinedNameSuggestionsIntro,"Maybe you want one of the following:"
-undefinedNameTypeParameter,"The type parameter '%s is not defined."
+undefinedNameTypeParameter,"The type parameter %s is not defined."
```
Comment (saul) [inline src/fsharp/FSComp.txt]: Was this change meant to have been made?

Comment (forki) [inline src/fsharp/FSComp.txt]: Yes. We add the tick back at call site. This is needed for suggestions to work in the VS quick fix tool

**Comment by forki** [comment]
@vasily-kirichenko Ok that bug was really weird. nobody noticed for years ;-)

Anyways: 

image

**Comment by vasily-kirichenko** [comment]
It's possible to show only "clean" attribute name if we at "attribute application" position, as I do in Add Open fix.

**Comment by forki** [comment]
@vasily-kirichenko not sure we already have that info during that phase already. will take a look

**Comment by vasily-kirichenko** [comment]
@forki Yes, you should already know that the type is an attribute, but I suspect you don't know that it's not an attribute definition, but applying.

**Comment by vasily-kirichenko** [comment]
Ah. Got it. Then do it on the Code Fix side.

**Comment by forki** [comment]
yes another filter layer at code fix site might be possible. Maybe in later PR

**Comment by forki** [comment]
this is now ready to go in. I think it brings great improvements - just look at the test cases. Many many of the false postives are gone. I hope it's still in time for RTM.

**Comment by saul** [comment]
Agreed - this would be great to have in RTM. This and a few other currently open PRs make the IDE experience significantly better.

**Comment by cartermp** [comment]
@KevinRansom I think this one counts as a significant improvement and meets the bar for RTM.  It makes suggestions far nicer.

**Comment by vasily-kirichenko** [comment]
https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharp.Editing.VisualStudio/Navigation/GoToDefinitionFilter.fs#L71

image

Everything else work _very_ well. Great job!

**Comment by vasily-kirichenko** [comment]
Found another one https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharp.Editing.VisualStudio/Navigation/GoToDefinitionFilter.fs#L42

image

here `VSLanguageService` is much more relevant then `SVsImageService` (IMHO) and should be placed on top of it.

**Comment by forki** [comment]
Mhm interesting case. Will take a look and send another pr if I find something. Its basically needed that we add even more suggestions - so it's a bit the opposite of this one here

**Comment by forki** [comment]
I think the last one is a bit weird. It's not a typical typo to remove first charecter. So metric is a bit wrong => won't fix

**Comment by vasily-kirichenko** [comment]
Here `string` should go first:

image

**Comment by forki** [comment]
String vs string is not decidable for our case insensitive metric

**Comment by vasily-kirichenko** [comment]
Oh, a rather cool one

image

**Comment by vasily-kirichenko** [comment]
And

image

here `IProjectProvider` is not suggested.

**Comment by vasily-kirichenko** [comment]
No suggestion:

https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharp.Editing.VisualStudio/Navigation/NoPeekDefinitionFilterProvider.fs#L95

image

The tricky thing about this is that `VisualStudioVersion` is both a module and a type:

image

**Pull request merged**